include/xtl/xthreadpool.hpp

```diff
@@ -0,0 +1,286 @@
+                            int(std::thread::hardware_concurrency()) / 2 : 
+                            int(std::thread::hardware_concurrency());
+        #endif
+    }
+
+    int numThreads_;
+};
+
+/********************************************************/
+/*                                                      */
+/*                      xthreadpool                      */
+/*                                                      */
+/********************************************************/
+
+    /**\brief Thread pool class to manage a set of parallel workers.
+
+        <b>\#include</b> \<xtl/xthreadpool.hpp\><br>
+        Namespace: xtl
+    */
+class xthreadpool
+{
+  public:
+
+    /** Create a thread pool from xparallel_options. The constructor just launches
+        the desired number of workers. If the number of threads is zero,
+        no workers are started, and all tasks will be executed in synchronously
+        in the present thread.
+     */
+    xthreadpool(const xparallel_options & options = xparallel_options())
+    :   options_(options),
+        stop(false)        
+    {
+        init();
+    }
+
+
+
+    /** Create a thread pool with n threads. The constructor just launches
+        the desired number of workers. If \arg n is <tt>xparallel_options::Auto</tt>,
+        the number of threads is determined by <tt>std::thread::hardware_concurrency()</tt>.
+        <tt>xparallel_options::Nice</tt> will create half as many threads.
+        If <tt>n = 0</tt>, no workers are started, and all tasks will be executed
+        synchronously in the present thread. If the preprocessor flag
+        <tt>xtl_SINGLE_THREADED</tt> is defined, the number of threads is always set
+        to zero (i.e. synchronous execution), regardless of the value of \arg n. This
+        is useful for debugging.
+     */
+    xthreadpool(const int n)
+    :   options_(n),
+        stop(false)
+    {
+        init();
+    }
+
+
+    /**
+     * The destructor joins all threads.
+     */
+    ~xthreadpool();
+
+    /**
+     * Enqueue a task that will be executed by the thread pool.
+     * The task result can be obtained using the get() function of the returned future.
+     * If the task throws an exception, it will be raised on the call to get().
+     */
+    template<class F>
+    auto enqueue(F&& f) -> std::future<decltype(f(0))>;
+
+
+    /**
+     * Block until all tasks are finished.
+     */
+    void wait()
+    {
+        std::unique_lock<std::mutex> lock(queue_mutex);
+        finish_condition.wait(lock, [this](){ return tasks.empty() && (busy == 0); });
+    }
+
+    /**
+     * Return the number of worker threads.
+     */
+    size_t n_worker_threads() const
+    {
+        return workers.size();
+    }
+
+    size_t n_threads() const
+    {
+        return workers.size() == 0 ? std::size_t(1) : workers.size();
+    }
+
+private:
+
+    // helper function to init the thread pool
+    void init();
+
+    xparallel_options options_;
```

Comment:
Not sure that we need to keep the `xparallel_options` as a class member since the number of threads is available through `workers.size()`.